src/context/backward_tokens.rs

```diff
@@ -20,7 +21,7 @@ macro_rules! move_to_prev_sibling_or_go_to_parent_and_loop {
     ($self:expr) => {
         match $self.node.prev_sibling() {
             None => {
-                $self.node = $self.node.parent().unwrap();
+                $self.node = $self.node.parent_(&$self.node_parent_provider);
```

Comment:
I went through and updated all usages of `Node::parent()` in this code-base to use the cache-based API's instead (haven't gone through and done the same in eg the eslint-builtin plugin yet, but I'd think the "rule of thumb" would be to always use those instead)